sdk/python/kfp/containers/entrypoint.py

```diff
@@ -156,7 +158,80 @@ def get_artifact(self) -> artifact.Artifact:
       return result
 
 
-def main(**kwargs):
+def _write_output_metadata_file(
+    fn_res: Union[int, str, float, NamedTuple],
+    output_artifacts: Dict[str, artifact.Artifact],
+    output_metadata_path: str
+):
+  """Writes the output metadata file to the designated place."""
+  # If output_params is a singleton value, needs to transform it to a mapping.
+  output_parameters = {}
+  if isinstance(fn_res, (int, str, float)):
+    output_parameters['output'] = fn_res
+  else:
+    # When multiple outputs, we'll need to match each field to the output paths.
+    for idx, output_name in enumerate(fn_res._fields):
+      output_parameters[output_name] = fn_res[idx]
+
+  executor_output = entrypoint_utils.get_executor_output(
+      output_artifacts=output_artifacts,
+      output_params=output_parameters)
+
+  with open(output_metadata_path, 'w') as f:
+    f.write(json_format.MessageToJson(executor_output))
+
+  return executor_output
+
+
+def main(
+    executor_input_str: str,
+    function_name: str,
+    output_metadata_path: Optional[str] = None):
+  """Container entrypoint used by KFP Python function based component
+
+  executor_input_str: A serialized ExecutorInput proto message.
+  function_name: The name of the user-defined function.
+  output_metadata_path: A local path where the output metadata JSON file should
+    be written to.
+  """
+  executor_input = pipeline_spec_pb2.ExecutorInput()
+  json_format.Parse(text=executor_input_str, message=executor_input)
+  output_metadata_path = output_metadata_path or executor_input.outputs.output_file
+  parameter_dict = {}  # kwargs to be passed to UDF.
+  for name, input_param in executor_input.inputs.parameters.items():
+    parameter_dict[name] = entrypoint_utils.get_python_value(input_param)
+
+  for name, input_artifacts in executor_input.inputs.artifacts.items():
+    parameter_dict[name] = artifact.Artifact.get_from_runtime_artifact(
+        input_artifacts.artifacts[0])
```

Comment:
For my education, there's always only one artifact? if so, why does this need to be a collection structure?

Reply:
This is a future-proof design where we would like to support multiple artifacts per-channel. For all the use cases in KFP and most conventional TFX it is safe to assume there's always only one artifact. 

That said, there's some certain use cases, especially in continuous training setting where multiple artifacts might be bundled into one channel, see, for example https://github.com/tensorflow/tfx/pull/2397

Reply:
I see, thanks for the explanation and reference.